File: merge_sort.js
function mergesort(myarray) {
    if (myarray.length === 1) {   //base case and is sorted [90]
        return myarray;
    }
    let midpoint = Math.floor(myarray.length/2);
    let leftside = myarray.slice(0,midpoint);   //90,80,70,50  ... 90 80 ... 90
    let rightside = myarray.slice(midpoint,);  //77,88,10,20   ... 70 50 ,,,, 80
    let sortleft = mergesort(leftside);      //80, 90
    let sortright = mergesort(rightside);     //50, 70
    let imyarray = merge(sortleft, sortright);
    return imyarray;
}

function merge(leftsorted, rightsorted) {
    let mergearray = [];
    while ((leftsorted.length !== 0) || (rightsorted.length !== 0)) {
        if (leftsorted.length === 0) {     //left side is empty but right side has some sorted value left, it should just append to merge array
            mergearray = mergearray.concat(rightsorted);
            rightsorted = [];                //emptying out the array so it will exit the while loop
        }
        else if (rightsorted.length === 0) {    ///right side is empty so left side is already sorted, so just add to mergearray
            mergearray = mergearray.concat(leftsorted);
            leftsorted = [];                 //emptying out the array so it will exit the while loop otherwise  it will loop forever
        }
        else if (leftsorted[0] > rightsorted[0]) {  //compares first value, right is smaller, add to mergearray
            mergearray.push(rightsorted.shift());
        }
        else if (leftsorted[0] < rightsorted[0]) {   //compares first value, left is smaller and add to array
            mergearray.push(leftsorted.shift());
        }
    }
    
    return mergearray;
}

myinputarray = [90,80,70,50,77,88,10,20];

console.log(mergesort(myinputarray));

File: merge_sort.py
###splitting up the array til it becomes 1 which means it is sorted
def mergesort(myarray):
    if len(myarray) == 1:
        return myarray   #[20]

    midpoint = len(myarray)//2   #midpoint using integer divison so it does the Math.floor
    leftside = myarray[:midpoint]       #10 20 30 ... then 10    ...    ...20
    rightside = myarray[midpoint:]      #40 50 80 ... then 20 30 ... then  ...30
    sortleft = mergesort(leftside)      #so once the left side got sorted goes to next line
    sortright = mergesort(rightside)    # sort out the right side
    myarray = merge(sortleft,sortright)   #this is where the magic happens and starts combing the sorted list on each level
    return myarray

def merge(leftarray, rightarray):
    mergeresult = []
    while len(leftarray) != 0 or len(rightarray) !=0:   ###something exist on both sides
        if len(leftarray) == 0:
            mergeresult.extend(rightarray)     ## so if nothing on left side, then just take the remaining right sided which is sorted and add to result
            rightarray = []                     ## empty out the right side
        elif len(rightarray) == 0:
            mergeresult.extend(leftarray)
            leftarray = []
        elif leftarray[0] < rightarray[0]:
            mergeresult.append(leftarray.pop(0))       ###pop out the zero element to be appended to the merge result
        elif leftarray[0] > rightarray[0]:
            mergeresult.append(rightarray.pop(0))
    return mergeresult                                  #return the sorted list


myarray = [80,30,20,50,40,10, 5]

v = mergesort(myarray)

print(v)
